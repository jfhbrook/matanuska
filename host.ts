import { Injectable } from '@nestjs/common';

import { basename } from 'path';
import * as os from 'os';
import { spawnSync } from 'child_process';
import { stdin, stdout, stderr } from 'node:process';

import { Readable, Writable } from 'stream';

import { ErrorCode } from './errors';
import { BaseException, FileError } from './exceptions';
import { Exit, ExitCode } from './exit';
import { DefaultFormatter } from './format';

/**
 * A logging level.
 */
export enum Level {
  Debug = 0,
  Info = 1,
  Warn = 2,
  Error = 3,
}

/**
 * An interface that encapsulates platform specific behavior. This includes:
 *
 * - Standard IO streams
 * - "Ground floor" logging
 * - IO for files and custom channels
 * - Process management
 * - Networking and other ports
 */
export interface Host {
  //
  // Ground floor I/O. IO is exposed as stdio streams so that lower level
  // components (like readline) can use them directly.
  //

  /**
   * The standard input stream.
   */
  inputStream: Readable;

  /**
   * The standard output stream.
   */
  outputStream: Writable;

  /**
   * The standard error stream.
   */
  errorStream: Writable;

  //
  // Logging concerns. It's arguable these shouldn't be the Host's problem,
  // but it needs to work at the ground floor, so we do it here.
  //

  /**
   * The current logging level. Used to suppress debug, info and warning
   * messages.
   */
  level: Level;

  /**
   * Set the current logging level.
   *
   * @param level The new logging level.
   */
  setLevel(level: Level): void;

  /**
   * Write a value to the output channel, without a newline.
   *
   * @param value The value to write.
   */
  writeOut(value: any): void;
  /**
   * Write a value to the error channel, without a newline.
   *
   * @param value The value to write.
   */
  writeError(value: any): void;

  /**
   * Write a value to the output channel, with a newline.
   *
   * @param value The value to write.
   */
  writeLine(value: any): void;

  /**
   * Write a value to the error channel, without a newline.
   *
   * @param value The value to write.
   */
  writeErrorLine(value: any): void;

  /**
   * Write a value to the debug channel. If the log level is not inclusive of
   * Debug, no output is written.
   *
   * @param value The value to write.
   */
  writeDebug(value: any): void;

  /**
   * Write a value to the info channel. If the log level is not inclusive of
   * Info, no output is written.
   *
   * @param value The value to write.
   */
  writeInfo(value: any): void;

  /**
   * Write a value to the warn channel. If the log level is not inclusive of
   * Warn, no output is written.
   *
   * @param value The value to write.
   */
  writeWarn(value: any): void;

  /**
   * Write an Exception to the error channel.
   *
   * @param exception The exception to write.
   */
  writeException(value: any): void;

  /**
   * Write a value to a numbered channel. The standard channels are:
   *
   * 1 - Output
   * 2 - Error
   * 3 - Warn
   * 4 - Info
   * 5 - Debug
   *
   * Channels 1 and 2 correspond to stdout and stderr, respectively.
   * Channels 3-5 are treated as logging methods.
   *
   * In the future, custom channels may be supported as an abstraction over
   * file descriptors.
   *
   * @param channel The channel to write to.
   * @param value The value to write.
   */
  writeChannel(channel: number, value: any): void;

  /**
   * The OS's hostname.
   */
  hostname(): string;

  /**
   * The OS's tty (if available).
   */
  tty(): string | null;

  /**
   * The basename of Matanuska BASIC's entry point script. This is a decent
   * approximation for the shell command as invoked.
   */
  shell(): string;

  /**
   * The current date and time.
   */
  now(): Date;

  /**
   * The current user's uid.
   */
  uid(): number;

  /**
   * The current user's gid.
   */
  gid(): number;

  /**
   * The current user's username.
   */
  username(): string;

  /**
   * The current user's home directory.
   */
  homedir(): string;

  /**
   * The process's current working directory. This is not the same as the
   * shell's current working directory.
   */
  cwd(): string;

  exit(code: number): void;
}

/**
 * A host for a standard terminal console.
 */
@Injectable()
export class ConsoleHost implements Host {
  private formatter = new DefaultFormatter();
  inputStream: Readable;
  outputStream: Writable;
  errorStream: Writable;
  level: Level;

  private _tty: string | null | undefined = undefined;

  constructor() {
    this.inputStream = stdin;
    this.outputStream = stdout;
    this.errorStream = stderr;
  }

  setLevel(level: Level): void {
    this.level = level;
  }

  writeOut(value: any): void {
    this.outputStream.write(this.formatter.format(value));
  }

  writeError(value: any): void {
    this.errorStream.write(this.formatter.format(value));
  }

  writeLine(value: any): void {
    this.outputStream.write(`${this.formatter.format(value)}\n`);
  }

  writeErrorLine(value: any): void {
    this.errorStream.write(`${this.formatter.format(value)}\n`);
  }

  writeDebug(value: any): void {
    if (this.level <= Level.Debug) {
      this.errorStream.write(`DEBUG: ${this.formatter.format(value)}\n`);
    }
  }

  writeInfo(value: any): void {
    if (this.level <= Level.Info) {
      this.errorStream.write(`INFO: ${this.formatter.format(value)}\n`);
    }
  }

  writeWarn(value: any): void {
    if (this.level <= Level.Warn) {
      this.errorStream.write(`WARN: ${this.formatter.format(value)}\n`);
    }
  }

  writeException(value: any): void {
    let exc = value;
    if (
      typeof value === 'string' ||
      typeof value === 'number' ||
      typeof value === 'boolean'
    ) {
      exc = new BaseException(String(value), null);
    }

    this.errorStream.write(`${this.formatter.format(exc)}\n`);
  }

  writeChannel(channel: number, value: any): void {
    switch (channel) {
      case 1:
        this.writeOut(value);
        break;
      case 2:
        this.writeError(value);
        break;
      case 3:
        this.writeWarn(value);
        break;
      case 4:
        this.writeInfo(value);
        break;
      case 5:
        this.writeDebug(value);
        break;
      default:
        throw new FileError(
          `Unknown channel: ${channel}`,
          ErrorCode.NoEntity,
          ExitCode.IoError,
          [`#${channel}`],
          null,
        );
    }
  }

  //
  // OS and environment concerns.
  //

  hostname(): string {
    return os.hostname();
  }

  tty(): string | null {
    if (typeof this._tty === 'undefined') {
      try {
        // I'm running this synronously because I assume it will be a quick
        // process and that it will only need to run once (since a process's
        // TTY never changes).
        const { stdout } = spawnSync('tty');
        this._tty = stdout.toString().trim();
      } catch (err) {
        // I'm assuming that if this fails, it's because it ultimately doesn't
        // make sense to ascribe a TTY to the process. But it would be nice
        // to have tracing for this, in some capacity.
        this._tty = null;
      }
    }

    return this._tty;
  }

  // A best attempt at the raw arguments to the command (ie. $0 in Bash). In
  // Bash (or a C program), you would have this behavior:
  //
  // - `bash` -> $0 is `bash`
  // - `$(which bash)` -> $0 is `/usr/bin/bash`
  //
  // Node.js does a bunch of processing to process.argv, such that the first
  // two arguments are *always* `node` and the path to the script,
  // respectively.
  //
  // There isn't a good portable way to get at the "original" argv from
  // Node.js. There are a few approaches:
  //
  // 1. Write a wrapper script in Bash that reads $0 directly, puts it in an
  //    environment variable, and then execs node.
  // 2. Write a wrapper in C++ (or Rust) that does something similar, but
  //    embeds Node.
  //
  // This works, for now, by doing the former. See ./bin/matbas for details.
  // But the latter could be compelling later, especially if I decide to
  // use rollup to build a bundle - if I embed the source build and Node
  // itself, then I'll have a static binary with no assets.
  //
  // As a fallback, just grab the basename of process.argv[1]. In most cases
  // it will be incorrect, but it's better than nothing.
  shell(): string {
    if (process.env.__MATBAS_DOLLAR_ZERO) {
      return process.env.__MATBAS_DOLLAR_ZERO;
    }
    return basename(process.argv[1]);
  }

  // TODO: JavaScript Dates aren't very good. Is there a sensible replacement?
  // TODO: Can we control locale-awareness better?
  now(): Date {
    return new Date();
  }

  uid(): number {
    return os.userInfo().uid;
  }

  gid(): number {
    return os.userInfo().gid;
  }

  username(): string {
    return os.userInfo().username;
  }

  homedir(): string {
    return os.homedir();
  }

  cwd(): string {
    return process.cwd();
  }

  exit(exitCode: number): void {
    throw new Exit(exitCode);
  }
}
