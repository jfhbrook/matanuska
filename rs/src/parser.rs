// This file will have the parser code. These will take a Tokens collection
// type and return Expressions, Statements, etc.
//
// I'm going to find limited value in naively implementing yabasic's parser.
// It's a very big parser, and I just don't need to complete it in order to
// get a working program.
//
// There are two directions to go here:
//
// 1. Implement the non-parser parts from yabasic. This will give me the
//    infrastructure to actually run simple commands, and give me a better
//    idea of how a parsed BASIC program is structured anyway.
// 2. Go through yabasic's parser and/or what I did in s7bas, and break feature
//    sets into milestones. Then, just implement those milestones one at a
//    time. This may also be a good time to delete some tokenization code I
//    know I won't use.
//

use crate::ast::{Command, Expr, Instruction, Line};
use crate::tokens::{Ident, Token, Tokens};
use nom::{
    branch::alt,
    bytes::complete::take,
    combinator::{map, verify},
    multi::separated_list0,
    sequence::{preceded, separated_pair},
    IResult,
};

// Like "tag" but for tokens, heavily informed by monkey-rust's parser. Note
// that the contents of the token must also match, since it uses an equality
// check - if you just need to match on the variant regardless of the insides,
// you'll have to figure out something else.
fn token<'a>(tag: Token) -> impl FnMut(Tokens<'a>) -> IResult<Tokens<'a>, Tokens<'a>> {
    let tag = tag.clone();
    move |input: Tokens<'a>| -> IResult<Tokens<'a>, Tokens<'a>> {
        verify(take(1usize), |t: &Tokens| t.tokens[0].token == tag)(input)
    }
}

// TODO: I'm going to need to write a whole family of these parsers - that
// is, combinators which match the token's variant and then pull the insides
// out into a useful type for mapping purposes. A macro might help?
fn strsym(input: Tokens) -> IResult<Tokens, Ident> {
    unimplemented!("strsym");
}

fn rem(input: Tokens) -> IResult<Tokens, String> {
    unimplemented!("rem");
}

// TODO: we should be able to parse open/close statements which are missing
// the close - loop, endif, next, until, wend

// TODO: I would like this to return Tokens instead of Vec<LocatedToken>, but
// the reference implementation actually expects to *borrow* the vec. They
// allocate the Vec, lend it to create Tokens, then pass that into the parser
// inside a function - example here:
//
//     https://github.com/Rydgel/monkey-rust/blob/master/lib/parser/mod.rs#L331-L336

fn instruction(input: Tokens) -> IResult<Tokens, Instruction> {
    unimplemented!("instruction");
}

fn command(input: Tokens) -> IResult<Tokens, Command> {
    unimplemented!("command");
}

pub fn expr(input: Tokens) -> IResult<Tokens, Expr> {
    unimplemented!("expr");
}

// yabasic includes scanning code which detects and emits a special token to
// mark an implicit endif. if we are to implement similar logic, it will
// happen in the parser code.
/*
pub fn implicit_endif(input: Span) -> IResult<Span, Token> {
    alt((
        token(Token::DoubleLineEnding),
        token(Token::SingleLineEnding),
        rem,
    ))(input)
}
*/

// This is some code where I started trying to write an assignment parser. It's
// not particularly useful at this stage, but it shows a bit of how the
// parsers would likely compose.
/*
type Assignment = Expr;

fn assignment(input: Tokens) -> IResult<Tokens, Assignment> {
    alt((
        string_assignment,
        preceded(token(Token::Let), string_assignment),
        number_assignment,
        preceded(token(Token::Let), number_assignment),
    ))(input)
}

fn string_assignment(input: Tokens) -> IResult<Tokens, Assignment> {
    alt((map(
        separated_pair(strsym, token(Token::Eq), string_expr),
        |(sym, expr)| {
            unimplemented!("add_command_with_sym_and_diag");
        },
    ),))(input)
}

fn number_assignment(input: Tokens) -> IResult<Tokens, Assignment> {
    unimplemented!("number_assignment");
}
*/
