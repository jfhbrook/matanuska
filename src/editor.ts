import { Line, LineNo } from './internal';

export class Editor {
  public program: Line[];

  constructor() {
    this.program = [];
  }

  insertLine(line: Line): void {}
  deleteLine(lineNo: LineNo): void {}
  nextLine(lineNo: LineNo): LineNo { return 100; };
}
