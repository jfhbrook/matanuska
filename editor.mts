import { Injectable, Inject } from '@nestjs/common';

import {
  ParseWarning,
  mergeParseErrors,
  removeFromParseError,
} from './exceptions.mjs';
import { Host } from './host.mjs';
import { Token } from './tokens.mjs';

import { Line, Program } from './ast/index.js';
import { Source } from './ast/source.js';
import {
  ExprVisitor,
  Unary,
  Binary,
  Logical,
  Group,
  Variable,
  IntLiteral,
  RealLiteral,
  BoolLiteral,
  StringLiteral,
  PromptLiteral,
  NilLiteral,
} from './ast/expr.js';
import {
  Instr,
  InstrVisitor,
  Rem,
  Let,
  Assign,
  Expression,
  Print,
  Exit,
  End,
  New,
  Load,
  List,
  Renum,
  Run,
  Save,
  ShortIf,
  If,
  Else,
  ElseIf,
  EndIf,
} from './ast/instr.js';

type LineNo = number;

interface Index {
  i: number;
  lineNo: LineNo | null;
  match: boolean;
}

// An index corresponding to inserting the line at the very front
const HEAD: number = -1;

type Numbering = {
  to: number;
  toStr: string;
  shift: number;
  source: Source;
};

type Renumbering = Record<LineNo, Numbering>;

export enum Justify {
  Left,
  Right,
}

class InstrShifter implements InstrVisitor<void>, ExprVisitor<void> {
  constructor(public shift: number) {}

  private shiftInstr(instr: Instr): void {
    instr.offsetStart += this.shift;
    instr.offsetEnd += this.shift;
  }

  private shiftToken(tok: Token): void {
    tok.offsetStart += this.shift;
    tok.offsetEnd += this.shift;
  }

  visitRemInstr(rem: Rem): void {
    this.shiftInstr(rem);
  }

  visitLetInstr(let_: Let): void {
    this.shiftInstr(let_);
    let_.variable.accept(this);
  }

  visitAssignInstr(assign: Assign): void {
    this.shiftInstr(assign);
    assign.variable.accept(this);
  }

  visitExpressionInstr(expr: Expression): void {
    this.shiftInstr(expr);
    expr.expression.accept(this);
  }

  visitPrintInstr(print: Print): void {
    this.shiftInstr(print);
    print.expression.accept(this);
  }

  visitExitInstr(exit: Exit): void {
    this.shiftInstr(exit);
    if (exit.expression) {
      exit.expression.accept(this);
    }
  }

  visitEndInstr(end: End): void {
    this.shiftInstr(end);
  }

  visitNewInstr(new_: New): void {
    this.shiftInstr(new_);
    if (new_.filename) {
      new_.filename.accept(this);
    }
  }

  visitLoadInstr(load: Load): void {
    this.shiftInstr(load);
    if (load.filename) {
      load.filename.accept(this);
    }
  }

  visitListInstr(list: List): void {
    this.shiftInstr(list);
  }

  visitRenumInstr(renum: Renum): void {
    this.shiftInstr(renum);
  }

  visitRunInstr(run: Run): void {
    this.shiftInstr(run);
  }

  visitSaveInstr(save: Save): void {
    this.shiftInstr(save);
    if (save.filename) {
      save.filename.accept(this);
    }
  }

  visitShortIfInstr(shortIf: ShortIf): void {
    this.shiftInstr(shortIf);
    shortIf.condition.accept(this);
    for (const instr of shortIf.then) {
      instr.accept(this);
    }
    for (const instr of shortIf.else_) {
      instr.accept(this);
    }
  }

  visitIfInstr(if_: If): void {
    this.shiftInstr(if_);
    if_.condition.accept(this);
  }

  visitElseInstr(else_: Else): void {
    this.shiftInstr(else_);
  }

  visitElseIfInstr(elseIf: ElseIf): void {
    this.shiftInstr(elseIf);
    elseIf.condition.accept(this);
  }

  visitEndIfInstr(endIf: EndIf): void {
    this.shiftInstr(endIf);
  }

  visitUnaryExpr(unary: Unary): void {
    unary.expr.accept(this);
  }

  visitBinaryExpr(binary: Binary): void {
    binary.left.accept(this);
    binary.right.accept(this);
  }

  visitLogicalExpr(logical: Logical): void {
    logical.left.accept(this);
    logical.right.accept(this);
  }

  visitGroupExpr(group: Group): void {
    group.expr.accept(this);
  }

  visitVariableExpr(variable: Variable): void {
    this.shiftToken(variable.ident);
  }

  visitIntLiteralExpr(_int: IntLiteral): void {}
  visitRealLiteralExpr(_real: RealLiteral): void {}
  visitBoolLiteralExpr(_bool: BoolLiteral): void {}
  visitStringLiteralExpr(_str: StringLiteral): void {}
  visitPromptLiteralExpr(_prompt: PromptLiteral): void {}
  visitNilLiteralExpr(_nil: NilLiteral): void {}
}

@Injectable()
export class Editor {
  public program: Program;
  public warning: ParseWarning | null;
  public justify: Justify = Justify.Left;

  constructor(@Inject('Host') public host: Host) {
    this.program = new Program('untitled.bas', []);
    this.warning = null;
  }

  get filename(): string {
    return this.host.relativePath('.', this.program.filename);
  }

  set filename(filename: string) {
    this.program.filename = this.host.resolvePath(filename);
  }

  /**
   * Initialize editor state.
   *
   * @param program A program.
   * @param warning An associated warning, if any.
   */
  init(program: Program, warning: ParseWarning | null): void {
    this.program = program;
    this.warning = warning;
  }

  /**
   * Reset the editor.
   */
  reset(): void {
    this.program = new Program('untitled.bas', []);
    this.warning = null;
  }

  /**
   * List the current program.
   *
   * @returns The source code for the program.
   */
  list(): string {
    return this.program.lines
      .map((l) => {
        if (
          l.instructions.length === 1 &&
          l.instructions[0] instanceof Rem &&
          l.instructions[0].remark === ''
        ) {
          return `${l.lineNo}`;
        }
        return l.source;
      })
      .join('\n');
  }

  /**
   * Renumber the current program.
   */
  renum(): void {
    const renumbering: Renumbering = {};

    // Used to calculate padding
    let maxToLength: number = 0;

    for (let i = 0; i < this.program.lines.length; i++) {
      const line = this.program.lines[i];

      // Calculate new line number
      const from = line.lineNo;
      const to = (i + 1) * 10;
      const toStr = String(to);
      renumbering[from] = {
        to,
        toStr,
        shift: 0,
        source: line.source,
      };

      // Track max length
      maxToLength = Math.max(maxToLength, toStr.length);
    }

    for (const line of this.program.lines) {
      const from = line.lineNo;
      const { to, toStr } = renumbering[from];

      // Justification
      const leftPadding =
        this.justify === Justify.Left
          ? ''
          : ' '.repeat(maxToLength - toStr.length);
      const rightPadding =
        this.justify === Justify.Left
          ? ' '.repeat(maxToLength - toStr.length + 1)
          : ' ';
      const fromWidth = line.source.prefix.length;
      const toWidth = leftPadding.length + toStr.length + rightPadding.length;
      const shift = toWidth - fromWidth;

      // Renumber the line
      line.lineNo = to;
      line.source.leadingWs = leftPadding;
      line.source.lineNo = toStr;
      line.source.separatingWs = rightPadding;

      // Shift the instructions
      for (const instr of line.instructions) {
        instr.accept(new InstrShifter(shift));
      }

      // Store for shifting the warnings
      renumbering[from].shift = shift;
    }

    // Renumber/format warnings
    if (this.warning) {
      for (let i = 0; i < this.warning.warnings.length; i++) {
        const warning = this.warning.warnings[i];
        const { to, shift, source } = renumbering[warning.lineNo];
        if (to) {
          warning.lineNo = to;
          // TODO: If the warning's source is known to be the same object
          // as line.source, this assignment is unnecessary. But I'd need to
          // look and confirm.
          warning.source = source;
          warning.offsetStart += shift;
          warning.offsetEnd += shift;
        }
      }
    }
  }

  // A binary search to find the index containing a lineNo. If there's no
  // exact match, return the index just prior.
  private findLineIndex(lineNo: number): Index {
    let lower = 0;
    let upper = this.program.lines.length - 1;

    if (!this.program.lines.length) {
      return { i: HEAD, lineNo: null, match: false };
    }

    let lowerLine = this.program.lines[lower];
    let upperLine = this.program.lines[upper];

    if (lowerLine.lineNo > lineNo) {
      return {
        i: HEAD,
        lineNo: null,
        match: false,
      };
    }

    if (lowerLine.lineNo === lineNo) {
      return {
        i: lower,
        lineNo: lowerLine.lineNo,
        match: true,
      };
    }

    if (upperLine.lineNo <= lineNo) {
      return {
        i: upper,
        lineNo: upperLine.lineNo,
        match: upperLine.lineNo === lineNo,
      };
    }

    // At this point, we know that the line, if defined, *must* exist
    // *between* lower and upper.

    while (true) {
      // We've exhausted our options and don't have a match. Since lower and
      // upper are guaranteed to be below and above our line respectively,
      // we can just return lower as the prior line.
      if (upper - lower <= 1) {
        return {
          i: lower,
          lineNo: lowerLine.lineNo,
          match: false,
        };
      }

      const middle = Math.floor((lower + upper) / 2);
      const middleLine = this.program.lines[middle];

      // Exact match!
      if (middleLine.lineNo === lineNo) {
        return {
          i: middle,
          lineNo: middleLine.lineNo,
          match: true,
        };
      }

      // middle is a new bound, either lower or upper
      if (middleLine.lineNo > lineNo) {
        upper = middle;
        upperLine = middleLine;
      } else if (middleLine.lineNo < lineNo) {
        lower = middle;
        lowerLine = middleLine;
      }
    }
  }

  /**
   * Set a line and its corresponding warnings.
   *
   * If a line has no instructions, delete the corresponding line. Otherwise,
   * insert or update the line based on lineNo.
   */
  setLine(line: Line, warning: ParseWarning | null): void {
    const { i, match } = this.findLineIndex(line.lineNo);

    if (!line.instructions.length) {
      if (match) {
        this.program.lines.splice(i, 1);
        this.warning = removeFromParseError(
          this.warning,
          'lineNo',
          line.lineNo,
        );
      }
      return;
    }

    if (match) {
      this.program.lines[i] = line;
      if (warning) {
        this.warning = mergeParseErrors([this.warning, warning]);
      }
      return;
    }

    this.program.lines.splice(i + 1, 0, line);
  }
}
