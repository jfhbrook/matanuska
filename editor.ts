import { Injectable, Inject } from '@nestjs/common';

import {
  ParseWarning,
  mergeParseErrors,
  removeFromParseError,
} from './exceptions';
import { Host } from './host';
import { Line, Program } from './ast';
import { Rem } from './ast/instr';

interface Index {
  i: number;
  lineNo: number | null;
  match: boolean;
}

// An index corresponding to inserting the line at the very front
const HEAD: number = -1;

@Injectable()
export class Editor {
  public program: Program;
  public warning: ParseWarning | null;

  constructor(@Inject('Host') public host: Host) {
    this.program = new Program('untitled.bas', []);
    this.warning = null;
  }

  get filename(): string {
    return this.host.relativePath('.', this.program.filename);
  }

  set filename(filename: string) {
    this.program.filename = this.host.resolvePath(filename);
  }

  /**
   * Initialize editor state.
   *
   * @param program A program.
   * @param warning An associated warning, if any.
   */
  init(program: Program, warning: ParseWarning | null): void {
    this.program = program;
    this.warning = warning;
  }

  /**
   * Reset the editor.
   */
  reset(): void {
    this.program = new Program('untitled.bas', []);
    this.warning = null;
  }

  /**
   * List the current program.
   *
   * @returns The source code for the program.
   */
  list(): string {
    return this.program.lines
      .map((l) => {
        if (
          l.instructions.length === 1 &&
          l.instructions[0] instanceof Rem &&
          l.instructions[0].remark === ''
        ) {
          return `${l.lineNo}`;
        }
        return l.source;
      })
      .join('\n');
  }

  /**
   * Renumber the current program.
   */
  renum(): void {
    const renumbering: Record<number, number> = {};

    for (let i = 0; i < this.program.lines.length; i++) {
      const line = this.program.lines[i];
      const from = line.lineNo;
      const to = (i + 1) * 10;
      line.lineNo = to;
      line.source.lineNo = String(to);
      renumbering[from] = to;
    }

    if (this.warning) {
      for (let i = 0; i < this.warning.warnings.length; i++) {
        const warning = this.warning.warnings[i];
        const to = renumbering[warning.lineNo];
        if (to) {
          warning.lineNo = to;
        }
      }
    }
  }

  // A binary search to find the index containing a lineNo. If there's no
  // exact match, return the index just prior.
  private findLineIndex(lineNo: number): Index {
    let lower = 0;
    let upper = this.program.lines.length - 1;

    if (!this.program.lines.length) {
      return { i: HEAD, lineNo: null, match: false };
    }

    let lowerLine = this.program.lines[lower];
    let upperLine = this.program.lines[upper];

    if (lowerLine.lineNo > lineNo) {
      return {
        i: HEAD,
        lineNo: null,
        match: false,
      };
    }

    if (lowerLine.lineNo === lineNo) {
      return {
        i: lower,
        lineNo: lowerLine.lineNo,
        match: true,
      };
    }

    if (upperLine.lineNo <= lineNo) {
      return {
        i: upper,
        lineNo: upperLine.lineNo,
        match: upperLine.lineNo === lineNo,
      };
    }

    // At this point, we know that the line, if defined, *must* exist
    // *between* lower and upper.

    while (true) {
      // We've exhausted our options and don't have a match. Since lower and
      // upper are guaranteed to be below and above our line respectively,
      // we can just return lower as the prior line.
      if (upper - lower <= 1) {
        return {
          i: lower,
          lineNo: lowerLine.lineNo,
          match: false,
        };
      }

      const middle = Math.floor((lower + upper) / 2);
      const middleLine = this.program.lines[middle];

      // Exact match!
      if (middleLine.lineNo === lineNo) {
        return {
          i: middle,
          lineNo: middleLine.lineNo,
          match: true,
        };
      }

      // middle is a new bound, either lower or upper
      if (middleLine.lineNo > lineNo) {
        upper = middle;
        upperLine = middleLine;
      } else if (middleLine.lineNo < lineNo) {
        lower = middle;
        lowerLine = middleLine;
      }
    }
  }

  /**
   * Set a line and its corresponding warnings.
   *
   * If a line has no instructions, delete the corresponding line. Otherwise,
   * insert or update the line based on lineNo.
   */
  setLine(line: Line, warning: ParseWarning | null): void {
    const { i, match } = this.findLineIndex(line.lineNo);

    if (!line.instructions.length) {
      if (match) {
        this.program.lines.splice(i, 1);
        this.warning = removeFromParseError(
          this.warning,
          'lineNo',
          line.lineNo,
        );
      }
      return;
    }

    if (match) {
      this.program.lines[i] = line;
      if (warning) {
        this.warning = mergeParseErrors([this.warning, warning]);
      }
      return;
    }

    this.program.lines.splice(i + 1, 0, line);
  }
}
