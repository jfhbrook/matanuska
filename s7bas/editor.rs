use std::fs::File;
use std::io::{BufRead, BufReader, Write};

use tracing::debug;

use crate::exception::Exception;
use crate::host::Host;
use crate::parse;
use crate::script::Script;
use crate::value::Value;

#[derive(Debug)]
pub struct Editor {
    script: Script,
}

impl Editor {
    pub fn new() -> Self {
        Editor {
            script: Script::new(),
        }
    }

    pub fn script(&self) -> &Script {
        &self.script
    }

    pub fn insert(&mut self, line_no: i64, line: String) {
        self.script.insert(line_no, line);
    }

    // TODO: right now I'm delegating exec to the Instruction impl, which
    // for Reasons requires that there are no generic methods
    // fn load_script<P: AsRef<Path> + std::fmt::Debug>(&mut self, filename: P) -> Result<()> {
    pub fn load_script(&mut self, filename: &String) -> Result<(), Exception> {
        let file = File::open(filename)?;
        let mut script = Script::new();

        for line in BufReader::new(file).lines() {
            let line = line?;

            let (line_no, line) = parse::input(line.as_str())?;

            if let Some(line_no) = line_no {
                script.insert(line_no, line);
            } else {
                debug!("non-numbered line: {}", line);
            }
        }

        self.script = script;

        Ok(())
    }

    // fn save_script<P: AsRef<Path> + std::fmt::Debug>(&mut self, filename: P) -> Result<()> {
    pub fn save_script(&mut self, filename: &String) -> Result<(), Exception> {
        let mut file = File::create(filename)?;
        for (line_no, line) in self.script.lines().iter() {
            file.write_all(format!("{} {}", line_no, line).as_bytes())?;
        }
        Ok(())
    }

    pub fn list<H>(&mut self, host: &H)
    where
        H: Host,
    {
        for (line_no, line) in self.script.lines().iter() {
            host.write_out(&Value::String(String::from(format!(
                "{} {}",
                line_no, line
            ))));
        }
    }
}
