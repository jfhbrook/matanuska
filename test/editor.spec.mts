import { describe, expect, test } from 'vitest';
import { discuss } from '@jfhbrook/swears';

import { Editor, Justify } from '../editor.mjs';

import { Line } from '../ast/index.mjs';

import { parseInput } from './helpers/parser.mjs';
import { MockConsoleHost } from './helpers/host.mjs';

type InsertFn = (source: string) => void;

const topic = discuss(async (): Promise<[Editor, InsertFn]> => {
  const editor = new Editor(new MockConsoleHost());
  function insert(source: string) {
    const [result] = parseInput(source);
    expect(result.input.length).toBe(1);
    const line = result.input[0];
    expect(line).toBeInstanceOf(Line);
    editor.setLine(line as Line, null);
  }

  return [editor, insert];
});

test('editor inserts', async () => {
  await topic.swear(async ([editor, insert]) => {
    insert('10 print "hello"');
    insert('30 print "world"');
    insert('20 print "hey"');
    insert('15 rem');
    insert('20');

    expect(editor.list()).toMatchSnapshot();
  });
});

// TODO: Test that editor.program has been shifted as expected. This will
// involve a full traversal of editor.program, twice, in order to calculate
// how each element shifted. Remember, expressions and tokens inside the
// program are also shifted!
//
// TODO: Test that editor.warnings has been shifted as expected. This will
// involve scanning editor.warnings, twice, in order to calcualte how each
// element shifted. It will also involve constructing lines which create
// warnings!
describe('editor renum', async () => {
  test('mixed double/triple to all double', async () => {
    await topic.swear(async ([editor, insert]) => {
      insert('10 print "foo"');
      insert('50 print "foo"');
      insert('100 print "foo"');
      insert('150 print "foo"');
      insert('200 print "foo"');
      insert('250 print "foo"');
      insert('300 print "foo"');
      insert('350 print "foo"');
      insert('400 print "foo"');

      editor.renum();

      expect(editor.list()).toMatchSnapshot();
    });
  });

  test('left justified to left justified', async () => {
    await topic.swear(async ([editor, insert]) => {
      insert('10  print "foo"');
      insert('50  print "foo"');
      insert('100 print "foo"');
      insert('150 print "foo"');
      insert('200 print "foo"');
      insert('250 print "foo"');
      insert('300 print "foo"');
      insert('350 print "foo"');
      insert('400 print "foo"');
      insert('450 print "foo"');

      editor.renum();

      expect(editor.list()).toMatchSnapshot();
    });
  });

  test('left justified to right justified', async () => {
    await topic.swear(async ([editor, insert]) => {
      editor.justify = Justify.Right;

      insert('10  print "foo"');
      insert('50  print "foo"');
      insert('100 print "foo"');
      insert('150 print "foo"');
      insert('200 print "foo"');
      insert('250 print "foo"');
      insert('300 print "foo"');
      insert('350 print "foo"');
      insert('400 print "foo"');
      insert('450 print "foo"');

      editor.renum();

      expect(editor.list()).toMatchSnapshot();
    });
  });

  test('right justified to left justified', async () => {
    await topic.swear(async ([editor, insert]) => {
      insert(' 10 print "foo"');
      insert(' 50 print "foo"');
      insert('100 print "foo"');
      insert('150 print "foo"');
      insert('200 print "foo"');
      insert('250 print "foo"');
      insert('300 print "foo"');
      insert('350 print "foo"');
      insert('400 print "foo"');
      insert('450 print "foo"');

      editor.renum();

      expect(editor.list()).toMatchSnapshot();
    });
  });
});
