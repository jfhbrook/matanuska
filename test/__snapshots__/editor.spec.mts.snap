// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`editor inserts 1`] = `
"10 print "hello"
15
30 print "world""
`;

exports[`editor renum > left justified to left justified 1`] = `
"10  print "foo"
20  print "foo"
30  print "foo"
40  print "foo"
50  print "foo"
60  print "foo"
70  print "foo"
80  print "foo"
90  print "foo"
100 print "foo""
`;

exports[`editor renum > left justified to right justified 1`] = `
" 10 print "foo"
 20 print "foo"
 30 print "foo"
 40 print "foo"
 50 print "foo"
 60 print "foo"
 70 print "foo"
 80 print "foo"
 90 print "foo"
100 print "foo""
`;

exports[`editor renum > mixed double/triple to all double 1`] = `
"10 print "foo"
20 print "foo"
30 print "foo"
40 print "foo"
50 print "foo"
60 print "foo"
70 print "foo"
80 print "foo"
90 print "foo""
`;

exports[`editor renum > right justified to left justified 1`] = `
"10  print "foo"
20  print "foo"
30  print "foo"
40  print "foo"
50  print "foo"
60  print "foo"
70  print "foo"
80  print "foo"
90  print "foo"
100 print "foo""
`;
